nautobot/extras/jobs.py

```diff
@@ -104,25 +105,190 @@ class Meta:
+        self.logger = get_task_logger(self.__module__)
 
-        self.request = None
         self.active_test = "main"
         self.failed = False
-        self._job_result = None
+        self.job_result = None
 
-        # Compile test methods and initialize results skeleton
-        self.test_methods = []
-
-        for method_name in dir(self):
-            if method_name.startswith("test_") and callable(getattr(self, method_name)):
-                self.test_methods.append(method_name)
+    def __call__(self, *args, **kwargs):
+        # Attempt to resolve serialized data back into original form by creating querysets or model instances
+        # If we fail to find any objects, we consider this a job execution error, and fail.
+        # This might happen when a job sits on the queue for a while (i.e. scheduled) and data has changed
+        # or it might be bad input from an API request, or manual execution.
+        try:
+            deserialized_kwargs = self.deserialize_data(kwargs)
+        except Exception:
+            stacktrace = traceback.format_exc()
+            self.log_failure(f"Error initializing job:\n```\n{stacktrace}\n```")
+        self.active_test = "run"
+        context_class = JobHookChangeContext if isinstance(self, JobHookReceiver) else JobChangeContext
+        change_context = context_class(user=self.job_result.user, context_detail=self.class_path)
+        with change_logging(change_context):
+            return self.run(*args, **deserialized_kwargs)
 
     def __str__(self):
         return str(self.name)
 
     # See https://github.com/PyCQA/pylint-django/issues/240 for why we have a pylint disable on each classproperty below
 
+    # TODO(jathan): Could be interesting for custom stuff when the Job is
+    # enabled in the database and then therefore registered in Celery
+    @classmethod
+    def on_bound(cls, app):
+        """Called when the task is bound to an app.
+
+        Note:
+            This class method can be defined to do additional actions when
+            the task class is bound to an app.
+        """
+
+    # TODO(jathan): Could be interesting for showing the Job's class path as the
+    # shadow name vs. the Celery task_name?
+    def shadow_name(self, args, kwargs, options):
+        """Override for custom task name in worker logs/monitoring.
+
+        Example:
+            .. code-block:: python
+
+                from celery.utils.imports import qualname
+
+                def shadow_name(task, args, kwargs, options):
+                    return qualname(args[0])
+
+                @app.task(shadow_name=shadow_name, serializer='pickle')
+                def apply_function_async(fun, *args, **kwargs):
+                    return fun(*args, **kwargs)
+
+        Arguments:
+            args (Tuple): Task positional arguments.
+            kwargs (Dict): Task keyword arguments.
+            options (Dict): Task execution options.
+        """
```

Comment:
Default implementation returns nothing? Should it instead return an empty string or something like that?

Reply:
From the [celery docs](https://docs.celeryq.dev/en/stable/userguide/tasks.html?highlight=before_start#before_start) on `before_start`: 
> The return value of this handler is ignored.